### PR TITLE
xplat: cover additional edge cases for interval caching

### PR DESCRIPTION
**Pull request by obastemur**
- test update: Opt out the case for system time backward jumps.
- improve implementation: CPU tick value is circular. Make sure we don't miss updating `now`

### CODE REVIEW

File: lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp [hunk 1]
```diff
@@ -64,28 +64,33 @@ namespace DateTime
...
 
-        if (diff >= INTERVAL_FOR_TICK_BACKUP) // max *ms to respond system time changes
+        if (diff <= data.previousDifference || diff >= INTERVAL_FOR_TICK_BACKUP) // max *ms to respond system time changes
         {
             double currentTime = GetSystemTimeREAL();
 
             // in case the system time wasn't updated backwards, and cache is still beyond...
             if (currentTime >= data.cacheSysTime && currentTime < data.cacheSysTime + INTERVAL_FOR_TICK_BACKUP)
             {
+                data.previousDifference = /*ULONGLONG*/-1; // Make sure next request won't use cache
```
Comment (jianchun) [inline lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp]: >/*ULONGLONG*/-1 [](start = 42, length = 15)

`previousDifference` type is actually `int`, why not ULONGLONG?

Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp]: I'm also surprised to see `int` there. Updated.

File: lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp [hunk 2]
```diff
@@ -64,28 +64,33 @@ namespace DateTime
...
 
-        if (diff >= INTERVAL_FOR_TICK_BACKUP) // max *ms to respond system time changes
+        if (diff <= data.previousDifference || diff >= INTERVAL_FOR_TICK_BACKUP) // max *ms to respond system time changes
         {
             double currentTime = GetSystemTimeREAL();
 
             // in case the system time wasn't updated backwards, and cache is still beyond...
             if (currentTime >= data.cacheSysTime && currentTime < data.cacheSysTime + INTERVAL_FOR_TICK_BACKUP)
             {
+                data.previousDifference = (ULONGLONG) -1; // Make sure next request won't use cache
```
Comment (jianchun) [inline lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp]: >// Make sure next request won't use cache [](start = 58, length = 41)

Just to understand this, what's the case that this time it didn't use cache but next time it uses cache? Is it because of tick cycle?

File: test/Date/xplatInterval.js [hunk 3]
```diff
@@ -8,13 +8,20 @@ if (WScript.Platform && WScript.Platform.OS != "win32") {
...
+
     for (var j = 0; j < 1e2; j++) {
         var pre_time = Date.now(), now;
         for(var i = 0; i < 1e6; i++) {
             now = Date.now();
             var diff = now - pre_time
 
-            if (diff < 0) throw new Error ("Timer interval has failed. diff < 0");
+            // INTERVAL_FOR_TICK_BACKUP = 5
+            // So, anything beyond 5ms is not subject to our testing here.
```
Comment (jianchun) [inline test/Date/xplatInterval.js]: What is the exact reason for current test failure? System time jumped back 5+ ms?

What if system jump back 5- ms?

File: lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp (hunk 2 above)
Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp]: we don't know when `next time` is. This method can be called much later right? so, previous state was -> `cache is expired` so, whatever the upcoming `tickCount` is, we don't care and update the real time.

File: test/Date/xplatInterval.js (hunk 3 above)
Comment (obastemur) [inline test/Date/xplatInterval.js]: The reason is parallel python jobs under tzset env. We do not control real system time. The one fails here the real system time backwards. We do not cache or control that.

Comment (obastemur) [inline test/Date/xplatInterval.js]: in case system time jumps back <= 5 ms, we most likely recover that internally.

Comment (jianchun) [inline test/Date/xplatInterval.js]: > in case system time jumps back <= 5 ms, we most likely recover that internally. (due to our cache is 5ms)

How does it recover? Suppose last time it returned `cacheSysTime + 2` using cache. Now 3 ticks later, system time jumps to `cacheSysTime - 1`. It'll return from line 89 and fail the test (now - prev == -3).

(Just try to understand this test stability.)

File: lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp (hunk 2 above)
Comment (jianchun) [inline lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp]: Sorry, I was trying to understand the purpose of adding `previousDifference`. What error scenario exactly does it try to fix?

Comment (obastemur) [inline lib/Runtime/PlatformAgnostic/Platform/Common/HiResTimer.cpp]: We previously didn't cover circularity. We cover now + use same veritable for forced recache

File: test/Date/xplatInterval.js (hunk 3 above)
Comment (obastemur) [inline test/Date/xplatInterval.js]: This scenario makes sense but isn't system timer granularity is somewhere around 10ms? I'm not saying we can't see that test is failing but we most likely don't and I prefer to see that test is failing on different environments instead of discarding the comparison whenever the system timer backwards.

**Comment by jianchun** [comment]
:shipit:

**Comment by obastemur** [comment]
@jianchun Thanks for the review
